src/ros_filter.cpp

```diff
@@ -2075,6 +2131,75 @@ namespace RobotLocalization
   }
 
   template<typename T>
+  void RosFilter<T>::applyPreviousState(MeasurementPtr measurement)
+  {
+    if (previousMeasurementStates_.count(measurement->topicName_) == 0)
+    {
+      RF_DEBUG("First differential integration for " << measurement->topicName_ << "\n");
+      measurement->measurement_ = filter_.getState();
+      measurement->covariance_ = filter_.getEstimateErrorCovariance();
+    }
+    else
+    {
+      const Eigen::VectorXd &prevState = previousMeasurementStates_[measurement->topicName_]->state_;
+      Eigen::VectorXd &curMeas = measurement->measurement_;
+
+      tf2::Transform prevStateTf;
+      tf2::Quaternion prevStateRot;
+      prevStateTf.setOrigin(
+        tf2::Vector3(prevState(StateMemberX), prevState(StateMemberY), prevState(StateMemberZ)));
+      prevStateRot.setRPY(prevState(StateMemberRoll), prevState(StateMemberPitch), prevState(StateMemberYaw));
+      prevStateTf.setRotation(prevStateRot);
+
+      RF_DEBUG("Previous filter state is:\n" << prevStateTf);
+
+      tf2::Transform curMeasTf;
+      tf2::Quaternion curMeasRot;
+      curMeasTf.setOrigin(
+        tf2::Vector3(curMeas(StateMemberX), curMeas(StateMemberY), curMeas(StateMemberZ)));
+      curMeasRot.setRPY(curMeas(StateMemberRoll), curMeas(StateMemberPitch), curMeas(StateMemberYaw));
+      curMeasTf.setRotation(curMeasRot);
+      curMeasTf = prevStateTf * curMeasTf;
+
+      curMeas(StateMemberX) = curMeasTf.getOrigin().getX();
+      curMeas(StateMemberY) = curMeasTf.getOrigin().getY();
+      curMeas(StateMemberZ) = curMeasTf.getOrigin().getZ();
+
+      RosFilterUtilities::quatToRPY(
+        curMeasTf.getRotation(), curMeas(StateMemberRoll), curMeas(StateMemberPitch), curMeas(StateMemberYaw));
+
```

Comment:
Note that I am not touching the covariance from the measurement here, so we'll have to adjust our input sources accordingly. scan-to-scan already behaves as I want it to, but the wheel odometry doesn't. This is fine, though, because I plan to fuse both velocity and pose from the wheel odometry, thereby ensuring that the behavior won't change too drastically w.r.t. wheel odometry.